cpp/src/Slice/Grammar.y

```diff
@@ -1653,14 +1653,29 @@ ICE_ENUM
 // ----------------------------------------------------------------------
 enumerator_list
 // ----------------------------------------------------------------------
-: enumerator ',' enumerator_list
+: meta_data enumerator ',' enumerator_list
 {
-    auto ens = dynamic_pointer_cast<EnumeratorListTok>($1);
-    ens->v.splice(ens->v.end(), dynamic_pointer_cast<EnumeratorListTok>($3)->v);
-    $$ = ens;
+    auto metadata = dynamic_pointer_cast<StringListTok>($1);
+    auto enumerator = dynamic_pointer_cast<Enumerator>($2);
+    if (enumerator && !metadata->v.empty())
+    {
+        enumerator->setMetaData(metadata->v);
+    }
+    auto enumeratorList = dynamic_pointer_cast<EnumeratorListTok>($4);
+    enumeratorList->v.push_front(enumerator);
+    $$ = enumeratorList;
 }
-| enumerator
+| meta_data enumerator
 {
+    auto metadata = dynamic_pointer_cast<StringListTok>($1);
+    auto enumerator = dynamic_pointer_cast<Enumerator>($2);
+    if (enumerator && !metadata->v.empty())
+    {
+        enumerator->setMetaData(metadata->v);
+    }
+    auto enumeratorList = make_shared<EnumeratorListTok>();
+    enumeratorList->v.push_front(enumerator);
+    $$ = enumeratorList;
```

Comment:
It's not totally clear to me why we have the %empty rule. Presumably the 'meta_data enumerator' rule above is sufficient to terminate the parsing of a non-empty enumeration.

According to:
https://doc.zeroc.com/ice/3.7/the-slice-language/user-defined-types/enumerations

> Slice does not permit empty enumerations.

Is this a bug and the parser actually allows empty enumerations?

In new Slice:
https://docs.icerpc.dev/slice2/language-guide/enum-types

> A checked enumeration must have at least one enumerator, while an unchecked enumeration may have no enumerator at all.

Reply:
> Is this a bug 

nope, ...

> and the parser actually allows empty enumerations?

... and yes!

----

You're correct, empty enums are disallowed in Slice, but we still have the logic to parse them correctly.
If we didn't have this rule, then an empty enum would cause a `syntax error: expected identifier, '[',  ...` to be emitted:
A) this error message isn't super helpful
B) and it isn't completely accurate, an empty enum is a _semantic_ error, not a _syntax_ error.
C) The bison compiler has no syntax-error recovery, so we only emit _true_ syntax errors as a last resort.

So, the idea is, we allow this at the parser level, and the place where we actually use this rule is where we do the check.
This way we can gracefully handle the error, and provide a more helpful error message. See:
https://github.com/zeroc-ice/ice/blob/6980c561cd479a6e5794193fe3e80b1eafaced83/cpp/src/Slice/Grammar.y#L1623-L1636

Reply:
Just as a note, if you ever look through the grammar you'll actually see alot of this.
'bugs' where the parser accepts completely bogus Slice!

For the reason that if we don't, we'll get Bison's default error message, and parsing halts immediately; both not great.
So, we allow the bogus Slice _for a little bit_ until we hit a place where we can safely reject it with a custom error message.